File: frontend/src/types/prediction.ts
export type PredictionLabel = "empty" | "human" | "object" | "unknown";

export interface PredictionRecord {
  macAddress: string;
  timestamp: string;
  label: PredictionLabel;
  confidence: number;
  distanceM: number;
  modelVersion: string;
  latencyMs: number;
  sequenceId?: number;
  rssi?: number;
  noiseFloor?: number;
  snr?: number;
}

export interface PredictionsResponse {
  predictions: PredictionRecord[];
}

export interface PaginatedPredictionsResponse {
  items: PredictionRecord[];
  total: number;
  page: number;
  pageSize: number;
}

File: backend/app/schemas/prediction.py
from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


PredictionLabel = Literal["empty", "human", "object", "unknown"]


class InferenceEnvelope(BaseModel):
    mac_address: str
    timestamp: datetime
    label: PredictionLabel
    confidence: float = Field(..., ge=0.0, le=1.0)
    distance_m: float = Field(..., ge=0.0)
    model_version: str
    latency_ms: float = Field(..., ge=0.0)


class PredictionRecord(InferenceEnvelope):
    sequence_id: Optional[int] = None
    rssi: Optional[float] = None
    noise_floor: Optional[float] = None
    snr: Optional[float] = None


class LatestPredictionsResponse(BaseModel):
    predictions: list[PredictionRecord]


class PaginatedPredictionsResponse(BaseModel):
    items: list[PredictionRecord]
    total: int
    page: int
    page_size: int
